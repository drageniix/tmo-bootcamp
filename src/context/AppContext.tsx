import React, { createContext } from "react";
import { BasicIngredient, Recipe } from "../api/recipe";
import { useRecipeSearch } from "../hooks/context";

export type AppContextType = {
  onSearchIngredients?: (input: string) => Promise<void>;
  onAddIngredient?: (chosenIngredient: BasicIngredient) => Promise<void>;
  onRemoveIngredient?: (
    id: number
  ) => (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => Promise<void>;
  ingredients: BasicIngredient[];
  searchedIngredients: BasicIngredient[];
  recipes: Recipe[];
};

export const defaultContext: AppContextType = {
  ingredients: [],
  recipes: [],
  searchedIngredients: [],
};

export const AppContext = createContext(defaultContext);

export const AppProvider = ({
  children,
}: {
  children: JSX.Element;
}): JSX.Element => {
  const appProps = useRecipeSearch();

  return <AppContext.Provider value={appProps}>{children}</AppContext.Provider>;
};
