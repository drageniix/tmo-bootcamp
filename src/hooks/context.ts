import { useState } from "react";
import {
  BasicIngredient,
  searchIngredients,
  searchRecipes,
} from "../api/recipe";
import { AppContextType, defaultContext } from "../context/AppContext";

export const useRecipeSearch = (): AppContextType => {
  const [recipes, setRecipes] = useState(defaultContext.recipes);
  const [ingredients, setIngredients] = useState(defaultContext.ingredients);
  const [searchedIngredients, setSearchedIngredients] = useState(
    defaultContext.searchedIngredients
  );

  const onSearchIngredients = async (input: string) => {
    const ingredients = await searchIngredients(input);
    setSearchedIngredients(ingredients);
  };

  const onAddIngredient = async (chosenIngredient: BasicIngredient) => {
    if (!ingredients.includes(chosenIngredient)) {
      const updatedIngredients = [...ingredients, chosenIngredient];
      const recipes = await searchRecipes(
        updatedIngredients.map((ingredient) => ingredient.name)
      );

      setIngredients(updatedIngredients);
      setRecipes(recipes);
    }
  };

  const onRemoveIngredient = (id: number) => async (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault();
    const updatedIngredients = ingredients.filter(
      (ingredient) => ingredient.id !== id
    );
    const recipes = await searchRecipes(
      updatedIngredients.map((ingredient) => ingredient.name)
    );

    setIngredients(updatedIngredients);
    setRecipes(recipes);
  };

  return {
    onSearchIngredients,
    onAddIngredient,
    onRemoveIngredient,
    searchedIngredients,
    ingredients,
    recipes,
  };
};
