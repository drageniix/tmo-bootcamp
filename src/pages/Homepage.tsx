import React from "react";
import SearchIngredients from "../components/form/SearchIngredients";
import IngredientList from "../components/ingredients/IngredientList";
import RecipeList from "../components/recipe/Recipes";
import { AppProvider } from "../context/AppContext";

export const Homepage = (): React.ReactElement => {
  return (
    <AppProvider>
      <main>
        <SearchIngredients />
        <IngredientList />
        <RecipeList />
      </main>
    </AppProvider>
  );
};

export default Homepage;
