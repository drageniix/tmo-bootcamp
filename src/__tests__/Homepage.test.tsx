/* eslint-disable jest/no-focused-tests */
import { AppContext, AppContextType, AppProvider } from "../context/AppContext";
import React from "react";
import { cleanup, render } from "@testing-library/react";
import Homepage from "../pages/Homepage";
import mockIngredients from "./mocks/ingredients.json";
import mockRecipes from "./mocks/recipes.json";

const renderForm = (context?: AppContextType) =>
  render(
    context ? (
      <AppContext.Provider value={context}>
        <Homepage />
      </AppContext.Provider>
    ) : (
      <AppProvider>
        <Homepage />
      </AppProvider>
    )
  );

describe("complete homepage", () => {
  afterEach(cleanup);

  it("should render full page without data", async () => {
    const { baseElement } = renderForm();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render full page with data", () => {
    const { baseElement } = renderForm({
      searchedIngredients: mockIngredients.results,
      ingredients: mockIngredients.results,
      recipes: mockRecipes,
    });

    expect(baseElement).toMatchSnapshot();
  });
});
