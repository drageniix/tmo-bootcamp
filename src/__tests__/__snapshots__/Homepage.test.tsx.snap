// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complete homepage should render full page with data 1`] = `
<body>
  <div>
    <main>
      <section
        class="ingredientSearch"
        data-testid="form"
      >
        <label
          class="ingredientSearch__label"
          for="downshift-2-input"
          id="downshift-3-label"
        >
          Choose an ingredient:
        </label>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="downshift-3-menu"
          class="ingredientSearch__actions"
          role="combobox"
        >
          <input
            aria-autocomplete="list"
            aria-controls="downshift-3-menu"
            aria-labelledby="downshift-3-label"
            autocomplete="off"
            class="ingredientSearch__actions--input"
            id="downshift-2-input"
            value=""
          />
          <button
            aria-label="add ingredient"
            class="ingredientSearch__actions--button"
            type="button"
          >
            +
          </button>
        </div>
        <ul
          aria-labelledby="downshift-3-label"
          class="ingredientSearch__list"
          id="downshift-3-menu"
          role="listbox"
        />
      </section>
      <section
        data-testid="ingredients"
      >
        <ul
          class="ingredientList"
        />
      </section>
      <ul
        class="recipeList"
        data-testid="recipes"
      />
    </main>
  </div>
</body>
`;

exports[`complete homepage should render full page without data 1`] = `
<body>
  <div>
    <main>
      <section
        class="ingredientSearch"
        data-testid="form"
      >
        <label
          class="ingredientSearch__label"
          for="downshift-0-input"
          id="downshift-1-label"
        >
          Choose an ingredient:
        </label>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="downshift-1-menu"
          class="ingredientSearch__actions"
          role="combobox"
        >
          <input
            aria-autocomplete="list"
            aria-controls="downshift-1-menu"
            aria-labelledby="downshift-1-label"
            autocomplete="off"
            class="ingredientSearch__actions--input"
            id="downshift-0-input"
            value=""
          />
          <button
            aria-label="add ingredient"
            class="ingredientSearch__actions--button"
            type="button"
          >
            +
          </button>
        </div>
        <ul
          aria-labelledby="downshift-1-label"
          class="ingredientSearch__list"
          id="downshift-1-menu"
          role="listbox"
        />
      </section>
      <section
        data-testid="ingredients"
      >
        <ul
          class="ingredientList"
        />
      </section>
      <ul
        class="recipeList"
        data-testid="recipes"
      />
    </main>
  </div>
</body>
`;
